File: app/src/main/java/ru/magflayer/spectrum/presentation/widget/PointView.java
package ru.magflayer.spectrum.presentation.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import ru.magflayer.spectrum.utils.AppUtils;

@SuppressWarnings("unused")
public class PointView extends View {

    private static final float RADIUS_DP = 4;
    private static final float STROKE_WIDTH_DP = 2;

    @SuppressWarnings("WeakerAccess")
    public interface OnPointChangeListener {
        void onPointChanged(float x, float y, int radius);
    }

    private Paint aimPaint;
    private float currentX;
    private float currentY;
    private float circleRadius;

    private RectF leftTopArc;
    private RectF leftBottomArc;
    private RectF rightTopArc;
    private RectF rightBottomArc;

    private boolean moveEnabled;
    private OnPointChangeListener onPointChangeListener;
    private boolean moveActive;

    public PointView(Context context) {
        this(context, null);
    }

    public PointView(Context context, AttributeSet attrs) {
        this(context, attrs, -1);
    }

    public PointView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        circleRadius = AppUtils.convertDpToPixel(RADIUS_DP, getContext());
        float circleStrokeWidth = AppUtils.convertDpToPixel(STROKE_WIDTH_DP, getContext());

        aimPaint = new Paint();
        aimPaint.setAntiAlias(true);
        aimPaint.setStrokeWidth(circleStrokeWidth);
        aimPaint.setStyle(Paint.Style.STROKE);

        leftTopArc = new RectF();
        leftBottomArc = new RectF();
        rightTopArc = new RectF();
        rightBottomArc = new RectF();
    }

    public void setOnPointChangeListener(OnPointChangeListener onPointChangeListener) {
        this.onPointChangeListener = onPointChangeListener;
    }

    public float getCurrentX() {
        return currentX;
    }

    public float getCurrentY() {
        return currentY;
    }

    public int getRadius() {
        return (int) circleRadius;
    }

    public void setMoveActive(boolean moveActive) {
        this.moveActive = moveActive;
    }

    public void setAimColor(int color) {
        aimPaint.setColor(color);
        invalidate();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        int parentWidth = MeasureSpec.getSize(widthMeasureSpec);
        int parentHeight = MeasureSpec.getSize(heightMeasureSpec);

        currentX = parentWidth / 2;
        currentY = parentHeight / 2;

        leftTopArc.set(currentX - 50, currentY - 50, currentX, currentY);
        leftBottomArc.set(currentX - 50, currentY, currentX, currentY + 50);
        rightTopArc.set(currentX, currentY - 50, currentX + 50, currentY);
        rightBottomArc.set(currentX, currentY, currentX + 50, currentY + 50);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        canvas.drawCircle(currentX, currentY, circleRadius, aimPaint);

        if (leftTopArc != null && leftBottomArc != null
                && rightTopArc != null && rightBottomArc != null) {
            canvas.drawArc(leftTopArc, 180, 90, false, aimPaint);
            canvas.drawArc(leftBottomArc, 180, -90, false, aimPaint);
            canvas.drawArc(rightTopArc, 0, -90, false, aimPaint);
            canvas.drawArc(rightBottomArc, 0, 90, false, aimPaint);
        }

        if (onPointChangeListener != null) {
            onPointChangeListener.onPointChanged(currentX, currentY, getRadius());
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        float x = event.getX();
        float y = event.getY();

        if (!moveActive) {
            return super.onTouchEvent(event);
        }

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                touchDown(x, y);
                break;
            case MotionEvent.ACTION_MOVE:
                touchMove(x, y);
                break;
            case MotionEvent.ACTION_UP:
                touchUp();
                break;
        }

        return true;
    }

    private void touchDown(float x, float y) {
        if ((x > currentX - circleRadius && x < currentX + circleRadius)
                && (y > currentY - circleRadius && y < currentY + circleRadius)) {
            moveEnabled = true;
            currentX = x;
            currentY = y;
            invalidate();
        }
    }

    private void touchMove(float x, float y) {
        if (moveEnabled) {
            currentX = x;
            currentY = y;
            invalidate();
        }
    }

    private void touchUp() {
        moveEnabled = false;
    }
}

File: app/src/main/java/ru/magflayer/spectrum/presentation/common/android/widget/PointView.kt
package ru.magflayer.spectrum.presentation.common.android.widget

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.util.DisplayMetrics
import android.view.MotionEvent
import android.view.View
import kotlin.math.roundToInt

class PointView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = -1,
) : View(context, attrs, defStyleAttr) {

    companion object {
        private const val RADIUS_DP = 4f
        private const val STROKE_WIDTH_DP = 2f
        private const val POINTER_SIZE = 50
    }

    private val aimPaint = Paint()
    var currentX: Float = 0.toFloat()
        private set
    var currentY: Float = 0.toFloat()
        private set

    private val leftTopArc = RectF()
    private val leftBottomArc = RectF()
    private val rightTopArc = RectF()
    private val rightBottomArc = RectF()

    private val radius = RADIUS_DP.dpToPx()
    private var circleRadius: Float = radius

    private var onPointChangeListener: OnPointChangeListener? = null
    private var moveEnabled: Boolean = false
    private var moveActive: Boolean = false

    interface OnPointChangeListener {
        fun onPointChanged(x: Float, y: Float, radius: Int)
    }

    init {
        val circleStrokeWidth = STROKE_WIDTH_DP.dpToPx()

        aimPaint.isAntiAlias = true
        aimPaint.strokeWidth = circleStrokeWidth
        aimPaint.style = Paint.Style.STROKE
    }

    fun setAimColor(color: Int) {
        aimPaint.color = color
        invalidate()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val parentWidth = MeasureSpec.getSize(widthMeasureSpec)
        val parentHeight = MeasureSpec.getSize(heightMeasureSpec)

        currentX = (parentWidth / 2).toFloat()
        currentY = (parentHeight / 2).toFloat()

        leftTopArc.set(currentX - POINTER_SIZE, currentY - POINTER_SIZE, currentX, currentY)
        leftBottomArc.set(currentX - POINTER_SIZE, currentY, currentX, currentY + POINTER_SIZE)
        rightTopArc.set(currentX, currentY - POINTER_SIZE, currentX + POINTER_SIZE, currentY)
        rightBottomArc.set(currentX, currentY, currentX + POINTER_SIZE, currentY + POINTER_SIZE)
    }

    override fun onDraw(canvas: Canvas) {
        canvas.drawCircle(currentX, currentY, circleRadius, aimPaint)

        canvas.drawArc(leftTopArc, 180f, 90f, false, aimPaint)
        canvas.drawArc(leftBottomArc, 180f, -90f, false, aimPaint)
        canvas.drawArc(rightTopArc, 0f, -90f, false, aimPaint)
        canvas.drawArc(rightBottomArc, 0f, 90f, false, aimPaint)

        onPointChangeListener?.onPointChanged(currentX, currentY, radius.roundToInt())
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        val x = event.x
        val y = event.y

        if (!moveActive) {
            return super.onTouchEvent(event)
        }

        return when (event.action) {
            MotionEvent.ACTION_DOWN -> touchDown(x, y)
            MotionEvent.ACTION_MOVE -> touchMove(x, y)
            MotionEvent.ACTION_UP -> touchUp()
            else -> false
        }
    }

    private fun touchDown(x: Float, y: Float): Boolean {
        if (x > currentX - circleRadius &&
            x < currentX + circleRadius &&
            y > currentY - circleRadius &&
            y < currentY + circleRadius
        ) {
            moveEnabled = true
            currentX = x
            currentY = y
            invalidate()
            return true
        }
        return false
    }

    private fun touchMove(x: Float, y: Float): Boolean {
        if (moveEnabled) {
            currentX = x
            currentY = y
            invalidate()
            return true
        }
        return false
    }

    private fun touchUp(): Boolean {
        moveEnabled = false
        return true
    }

    private fun Float.dpToPx(): Float {
        val resources = context.resources
        val metrics = resources.displayMetrics
        return this * (metrics.densityDpi.toFloat() / DisplayMetrics.DENSITY_DEFAULT)
    }
}
